### PR TITLE
MaxMin split

### PR DESCRIPTION
**Pull request by jakadam2**
## Changes

*Short description of changes*

## Checklist before requesting a review
- [x] Docstrings added/updated in public functions and classes
- [x] Tests added, reasonable test coverage (at least ~90%, `make test-coverage`)
- [x] Sphinx docs added/updated and render properly (`make docs` and see `docs/_build/index.html`)

### CODE REVIEW

**Comment by jakadam2** [comment]
documentation comments are still in progress but if you would have a piece of free time, I belive it is good moment for review of code

File: skfp/model_selection/splitters/maxmin_split.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    """
+    data_size = len(data)
+    train_size, test_size = validate_train_test_split_sizes(
+        train_size, test_size, data_size
+    )
+
+    molecules = ensure_mols(data)
+    fingerprints = [
+        AllChem.GetMorganFingerprintAsBitVect(mol, 2, 1024) for mol in molecules
+    ]
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:99]: 1. Add arguments `fingerprints` with default value None. If only they are not provided, calculate this
2. Use `ECFPFingerprint` instead of raw RDKit

File: skfp/model_selection/splitters/maxmin_split.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+    picker = MaxMinPicker()
+    test_idxs = picker.LazyPick(
+        distFunc=lambda i, j: 1
+        - DataStructs.DiceSimilarity(fingerprints[i], fingerprints[j]),
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:104]: Create as a separate variable

File: skfp/model_selection/splitters/maxmin_split.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    )
+
+    valid_idxs = picker.LazyPick(
+        distFunc=lambda i, j: 1
+        - DataStructs.DiceSimilarity(fingerprints[i], fingerprints[j]),
+        poolSize=data_size,
+        pickSize=test_size + valid_size,
+        firstPicks=test_idxs,
+        seed=random_state,
+    )
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:193]: Add a comment for `firstPicks` and how this works. This is smart and efficient, but not obvious

File: tests/model_selection/splitters/maxmin_split.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+from skfp.model_selection.splitters.maxmin_split import (
+    maxmin_train_test_split,
+    maxmin_train_valid_test_split,
+)
+
+
+@pytest.fixture
+def all_molecules() -> list[str]:
+    all_smiles: list[str] = [
```
Comment (j-adamczyk) [inline tests/model_selection/splitters/maxmin_split.py:15]: No need for additional typing here, just `return [...]`

**Review by j-adamczyk — CHANGES_REQUESTED** [review]

File: tests/model_selection/splitters/maxmin_split.py (hunk 4 above)
Comment (jakadam2) [reply tests/model_selection/splitters/maxmin_split.py]: done

File: skfp/model_selection/splitters/maxmin_split.py (hunk 3 above)
Comment (jakadam2) [reply skfp/model_selection/splitters/maxmin_split.py]: done

File: skfp/model_selection/splitters/maxmin_split.py (hunk 2 above)
Comment (jakadam2) [reply skfp/model_selection/splitters/maxmin_split.py]: done

File: skfp/model_selection/splitters/maxmin_split.py (hunk 1 above)
Comment (jakadam2) [reply skfp/model_selection/splitters/maxmin_split.py]: 1. Done
2. Lets talk about it on weekly, I thought rawRDKit with radious > 1, is like ECFPFingerprint

Comment (j-adamczyk) [reply skfp/model_selection/splitters/maxmin_split.py]: It is, so why not use it? We have our own classes for convenience, after all, and it's shorter:
```
fps = ECFPFingerprint().transform(molecules)
```
Also it uses more standard 2048 bits by default.

File: skfp/model_selection/splitters/maxmin_split.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+    tuple[Sequence, ...],
+    tuple[Sequence[int], Sequence[int]],
+]:
+    """
+    Split using maxmin algorithm of picking a subset of item from a pool.
+    Starting from random item of initial set, next is picked item with maximum
+    value for its minimum distance to molecules in the picked set (hence the MaxMin name),
+    calculating and recording the distances as required. This molecule is the most distant
+    one to those already picked so is transferred to the picked set
+
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:60]: Add a very short single line at the beginning with minimal description. It is rendered in tables in the docs. See fingerprints or filters for examples.

Description should follow the general style of scaffold / Butina (see PR) docs, see them for examples. Dot is missing from the last sentence.

File: skfp/model_selection/splitters/maxmin_split.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+    return_indices : bool, default=False
+        Whether the method should return the input object subsets, i.e. SMILES strings
+        or RDKit `Mol` objects, or only the indices of the subsets instead of the data.
+
+    Returns
+    ----------
+    subsets : tuple[list, list, ...]
+    Tuple with train-test subsets of provided arrays. First two are lists of SMILES strings or RDKit `Mol` objects,
+    depending on the input type. If `return_indices` is True, lists of indices are returned instead of actual data.
+    References
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:87]: Add empty line before "References"

File: skfp/model_selection/splitters/maxmin_split.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+    ----------
+    https://github.com/deepchem/deepchem
+    https://rdkit.org/docs/cppapi/classRDPickers_1_1MaxMinPicker.html
+    https://squonk.it/docs/cells/RDKit%20MaxMin%20Picker/
+    https://rdkit.blogspot.com/2017/11/revisting-maxminpicker.html
+
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:93]: Follow proper citation format. See fingerprints for examples how we cite papers and websites. Also use them in text if relevant, e.g. with `[1]_`. Also remove empty line at the end of the docstring

File: skfp/model_selection/splitters/maxmin_split.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+    """
+    data_size = len(data)
+    train_size, test_size = validate_train_test_split_sizes(
+        train_size, test_size, data_size
+    )
+
+    molecules = ensure_mols(data)
+
+    if fingerprints is None:
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:102]: I think we can remove `fingerprints` argument and just compute them here. We need specific RDKit objects anyway

File: skfp/model_selection/splitters/maxmin_split.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+        fingerprints = [
+            AllChem.GetMorganFingerprintAsBitVect(mol, radius=4, nBits=1024)
+            for mol in molecules
+        ]
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:106]: Radius 2, not 4. ECFP4 means diameter 4, so radius 2. Also use more standard 2048 bits. This can also be a one-liner when a generator is used. See Butina split PR

File: skfp/model_selection/splitters/maxmin_split.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+
+    picker = MaxMinPicker()
+    test_idxs = picker.LazyPick(
+        distFunc=create_dice_distance_function_from_fingerprints(fingerprints),
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:110]: You can use `LazyBitVectorPick` and avoid using custom distance function altogether

File: skfp/model_selection/splitters/maxmin_split.py [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+        poolSize=data_size,
+        pickSize=test_size,
+        seed=random_state,
+    )
+
+    train_idxs = list(set(range(data_size)) - set(test_idxs))
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:116]: Small thing, but move `train_idxs` to the line above and make empty line after it

File: tests/model_selection/splitters/maxmin_split.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+def additional_data() -> list[list[Union[str, int, bool]]]:
+    return [
+        ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"],
+        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
+        [True, False, True, False, True, False, True, False, True, False],
+    ]
+
+
+def test_maxmin_split_size(all_molecules):
+
```
Comment (j-adamczyk) [inline tests/model_selection/splitters/maxmin_split.py:39]: Remove empty lines from beginning of functions

File: tests/model_selection/splitters/maxmin_split.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+    ]
+
+
+@pytest.fixture
+def additional_data() -> list[list[Union[str, int, bool]]]:
+    return [
+        ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"],
+        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
+        [True, False, True, False, True, False, True, False, True, False],
+    ]
```
Comment (j-adamczyk) [inline tests/model_selection/splitters/maxmin_split.py:35]: Unused

File: tests/model_selection/splitters/maxmin_split.py [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+    )
+
+    assert (
+        train_set_1 != train_set_2
+        or valid_set_1 != valid_set_2
+        or test_set_1 != test_set_2
+    )
+
+
+def test_seed_diversity_train_test_split(all_molecules):
```
Comment (j-adamczyk) [inline tests/model_selection/splitters/maxmin_split.py:120]: As above

File: tests/model_selection/splitters/maxmin_split.py [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+    train_set_2, test_set_2 = maxmin_train_test_split(
+        all_molecules, train_size=0.7, test_size=0.3, random_state=0
+    )
+
+    assert train_set_1 == train_set_2
+    assert test_set_1 == test_set_2
+
+
+def test_seed_diversity_train_valid_test_split(all_molecules):
```
Comment (j-adamczyk) [inline tests/model_selection/splitters/maxmin_split.py:103]: I don't think this test is necessary. After all, random_state 0 and 1 can result in the same result, and it's not an error

File: tests/model_selection/splitters/maxmin_split.py [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+@pytest.fixture
+def additional_data() -> list[list[Union[str, int, bool]]]:
+    return [
+        ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"],
+        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
+        [True, False, True, False, True, False, True, False, True, False],
+    ]
+
+
+def test_maxmin_split_size(all_molecules):
```
Comment (j-adamczyk) [inline tests/model_selection/splitters/maxmin_split.py:38]: Change name to `train_test_split_size`, following scaffold split test naming. We want to know from the test name which function we are testing and for what behavior

File: tests/model_selection/splitters/maxmin_split.py [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+        all_molecules, train_size=0.7, test_size=0.3, random_state=1
+    )
+
+    assert train_set_1 != train_set_2 or test_set_1 != test_set_2
+
+
+def test_maxmin_train_test_split_return_molecules(all_molecules):
+    mols = [Chem.MolFromSmiles(smiles) for smiles in all_molecules]
+    train_set, test_set = maxmin_train_test_split(
+        mols, train_size=0.7, test_size=0.3, random_state=1, return_indices=False
```
Comment (j-adamczyk) [inline tests/model_selection/splitters/maxmin_split.py:136]: Why specify `return_indices=False` when it's false by default?

Comment (j-adamczyk) [reply tests/model_selection/splitters/maxmin_split.py:136]: Also don't use `random_state` in tests unless necessary. I would think that it's necessary to pass it here for determinism, when it's not

**Review by j-adamczyk — CHANGES_REQUESTED** [review]

File: tests/model_selection/splitters/maxmin_split.py (hunk 15 above)
Comment (jakadam2) [reply tests/model_selection/splitters/maxmin_split.py]: agree, fixed

File: skfp/model_selection/splitters/maxmin_split.py (hunk 9 above)
Comment (jakadam2) [reply skfp/model_selection/splitters/maxmin_split.py]: done

File: skfp/model_selection/splitters/maxmin_split.py (hunk 8 above)
Comment (jakadam2) [reply skfp/model_selection/splitters/maxmin_split.py]: done

File: tests/model_selection/splitters/maxmin_split.py (hunk 17 above)
Comment (jakadam2) [reply tests/model_selection/splitters/maxmin_split.py]: done

File: skfp/model_selection/splitters/maxmin_split.py (hunk 7 above)
Comment (jakadam2) [reply skfp/model_selection/splitters/maxmin_split.py]: I have tried to correct my docs, believe now it's better

**Comment by jakadam2** [comment]
Changes done, please review if You would have a while, thanks

File: skfp/model_selection/splitters/maxmin_split.py [hunk 18]
```diff
@@ -0,0 +1,311 @@
...
+    ],
+    tuple[Sequence, ...],
+    tuple[Sequence[int], Sequence[int]],
+]:
+    """
+    Split using maxmin algorithm.
+
+    MaxMinPicker is an efficient algorithm for picking a optimal subset of diverse compounds from a candidate pool.
+    The algorithm is described in Ashton, M. [2]_,
+    improved RDKit implementation was described by Roger Sayle at the 2017 RDKit user group meeting [3]_.[5]_
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:58]: 1. Why weird line break? Just continue the sentence
2. `implementation was created by Roger Sayle [3]_ [4]_ [5]_`

File: skfp/model_selection/splitters/maxmin_split.py [hunk 19]
```diff
@@ -0,0 +1,311 @@
...
+
+    Starting from random item of initial set of vectorized molecules by binary ECFP4 fingerprint (radius 2) with
+    2048 bits, next is picked item with maximum
+    value for its minimum distance to molecules in the picked set (hence the MaxMin name),
+    calculating and recording the distances as required. This molecule is the most distant
+    one to those already picked so is transferred to the picked set [4]_.
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:64]: This is a very long sentence and hard to read. Shorter:
```
It starts by vectorizing molecules with ECFP4 fingerprint. First test molecule is picked randomly. Each next one is selected to maximize the minimal distance to the already selected molecules (hence the MaxMin name) [4]_.
```

File: skfp/model_selection/splitters/maxmin_split.py [hunk 20]
```diff
@@ -0,0 +1,311 @@
...
+    random_state: int = 0,
+) -> Union[
+    tuple[
+        Sequence[Union[str, Mol]], Sequence[Union[str, Mol]], Sequence[Sequence[Any]]
+    ],
+    tuple[Sequence, ...],
+    tuple[Sequence[int], Sequence[int]],
+]:
+    """
+    Split using maxmin algorithm.
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:54]: MaxMin

File: skfp/model_selection/splitters/maxmin_split.py [hunk 21]
```diff
@@ -0,0 +1,311 @@
...
+    Returns
+    ----------
+    subsets : tuple[list, list, ...]
+    Tuple with train-test subsets of provided arrays. First two are lists of SMILES strings or RDKit `Mol` objects,
+    depending on the input type. If `return_indices` is True, lists of indices are returned instead of actual data.
+
+    References
+    ----------
+    .. [1] `MaxMain split implementation and its variants
+        https://github.com/deepchem/deepchem_`
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:96]: No need to quote DeepChem IMO

File: skfp/model_selection/splitters/maxmin_split.py [hunk 22]
```diff
@@ -0,0 +1,311 @@
...
+    .. [3] `Roger Sayle
+        "Improved RDKit implementation"
+        https://github.com/rdkit/UGM_2017/blob/master/Presentations/Sayle_RDKitDiversity_Berlin17.pdf_`
+
+    .. [4] `Tim Dudgeon
+        "Revisting the MaxMinPicker"
+        https://rdkit.org/docs/cppapi/classRDPickers_1_1MaxMinPicker.html_`
+
+    .. [5] `RDKit MaxMin Picker
+        https://squonk.it/docs/cells/RDKit%20MaxMin%20Picker_`
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:112]: Remember to put URLs in <>, e.g. `<https://...>`. See other docs for examples

File: skfp/model_selection/splitters/maxmin_split.py [hunk 23]
```diff
@@ -0,0 +1,311 @@
...
+    tuple[
+        Sequence[Union[str, Mol]], Sequence[Union[str, Mol]], Sequence[Sequence[Any]]
+    ],
+    tuple[Sequence, ...],
+    tuple[Sequence[int], Sequence[int]],
+]:
+    """
+    Split using maxmin algorithm.
+
+        MaxMinPicker is an efficient algorithm for picking a optimal subset of diverse compounds from a candidate pool.
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:190]: Unnecessary additional indentation. Also remember to introduce changes requested for train-test function here

File: skfp/model_selection/splitters/maxmin_split.py [hunk 24]
```diff
@@ -0,0 +1,311 @@
...
+
+    picker = MaxMinPicker()
+    test_idxs = picker.LazyBitVectorPick(
+        fps,
+        poolSize=data_size,
+        pickSize=test_size,
+        seed=random_state,
+    )
+
+    # firstPicks - initial state of picked set
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:273]: This picks both train and valid set, right? And you filter it below. It would be nice to write this explicitly here

**Review by j-adamczyk — CHANGES_REQUESTED** [review]

**Comment by jakadam2** [comment]
done

File: skfp/model_selection/splitters/maxmin_split.py [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+    ],
+    tuple[Sequence, ...],
+    tuple[Sequence[int], Sequence[int]],
+]:
+    """
+    Split using MaxMin algorithm.
+
+    MaxMinPicker is an efficient algorithm for picking a optimal subset of diverse compounds from a candidate pool.
+    The algorithm is described in Ashton, M. [1]_,implementation was created by Roger Sayle [3]_ [4]_ [5]_
+    at the 2017 RDKit user group meeting [2]_.[4]_
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:58]: 1. We generally don't mention papers explicitly. You can reword this as `The original algorithm was introduced in [1]_, but here we use an optimized implementation by Roger Sayle [2]_ [3]_ [4]_ [5]_`
2. Don't mention RDKit UGM, it's already in reference.
3. Don't put any references after the dot
4. There is no reference 5?

File: skfp/model_selection/splitters/maxmin_split.py [hunk 26]
```diff
@@ -0,0 +1,291 @@
...
+
+    It starts by vectorizing molecules with ECFP4 fingerprint. First test molecule is picked randomly. Each next one
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:60]: I think it should be "The first test molecule"?

File: skfp/model_selection/splitters/maxmin_split.py [hunk 27]
```diff
@@ -0,0 +1,291 @@
...
+    subsets : tuple[list, list, ...]
+    Tuple with train-test subsets of provided arrays. First two are lists of SMILES strings or RDKit `Mol` objects,
+    depending on the input type. If `return_indices` is True, lists of indices are returned instead of actual data.
+
+    References
+    ----------
+    .. [1] `Mark Ashton, John Barnardb, Florence Casset, Michael Charlton, Geoffrey Downsb,
+        Dominique Gorse, JohnHolliday, Roger Lahanac, Peter Willett
+        "Identification of Diverse Database Subsets using Property-Based and Fragment-Based Molecular Descriptions"
+        <https://onlinelibrary.wiley.com/doi/10.1002/qsar.200290002>_`
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:96]: Separate references with single empty line

File: skfp/model_selection/splitters/maxmin_split.py [hunk 28]
```diff
@@ -0,0 +1,291 @@
...
+
+    Returns
+    ----------
+    subsets : tuple[list, list, ...]
+    Tuple with train-test subsets of provided arrays. First two are lists of SMILES strings or RDKit `Mol` objects,
+    depending on the input type. If `return_indices` is True, lists of indices are returned instead of actual data.
+
+    References
+    ----------
+    .. [1] `Mark Ashton, John Barnardb, Florence Casset, Michael Charlton, Geoffrey Downsb,
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:93]: I think that "Mark Ashton et al." will be reasonable here

File: skfp/model_selection/splitters/maxmin_split.py [hunk 29]
```diff
@@ -0,0 +1,291 @@
...
+        Dominique Gorse, JohnHolliday, Roger Lahanac, Peter Willett
+        "Identification of Diverse Database Subsets using Property-Based and Fragment-Based Molecular Descriptions"
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:95]: Add journal reference after title: Quant. Struct.-Act. Relat., 21: 598-604

File: skfp/model_selection/splitters/maxmin_split.py [hunk 30]
```diff
@@ -0,0 +1,291 @@
...
+    return_indices: bool = False,
+    random_state: int = 0,
+) -> Union[
+    tuple[
+        Sequence[Union[str, Mol]], Sequence[Union[str, Mol]], Sequence[Sequence[Any]]
+    ],
+    tuple[Sequence, ...],
+    tuple[Sequence[int], Sequence[int]],
+]:
+    """
```
Comment (j-adamczyk) [inline skfp/model_selection/splitters/maxmin_split.py:178]: Implement formatting like for train-test split, both description comments and empty lines (e.g. between references)

**Review by j-adamczyk — CHANGES_REQUESTED** [review]

**Review by j-adamczyk — APPROVED** [review]
